src/backend/data.rs

```diff
@@ -0,0 +1,193 @@
+        let path = id.path();
+        if !fs.exists(&path, location) {
+            return Err(Error::NotFound);
+        }
+        let data = fs
+            .read::<SIZE>(&path, location)
+            .map_err(|_| Error::ReadFailed)?;
+        let mut data: Self =
+            trussed::cbor_deserialize(&data).map_err(|_| Error::DeserializationFailed)?;
+        data.id = id;
+        Ok(data)
+    }
+
+    pub fn save<S: Filestore>(&self, fs: &mut S, location: Location) -> Result<(), Error> {
+        let data = trussed::cbor_serialize_bytes::<_, SIZE>(self)
+            .map_err(|_| Error::SerializationFailed)?;
+        fs.write(&self.id.path(), location, &data)
+            .map_err(|_| Error::WriteFailed)
+    }
+
+    pub fn retries_left(&self) -> Option<u8> {
+        self.retries.map(|retries| retries.left)
+    }
+
+    pub fn is_blocked(&self) -> bool {
+        if let Some(retries) = self.retries {
+            retries.left == 0
+        } else {
+            false
+        }
+    }
+
+    pub fn write<S, F, R>(&mut self, fs: &mut S, location: Location, f: F) -> Result<R, Error>
+    where
+        S: Filestore,
+        F: Fn(&mut PinDataMut<'_>) -> R,
+    {
+        let mut data = PinDataMut::new(self);
+        let result = f(&mut data);
+        if data.modified {
+            self.save(fs, location)?;
+        }
+        Ok(result)
+    }
+}
+
+pub(crate) struct PinDataMut<'a> {
+    data: &'a mut PinData,
+    modified: bool,
+}
+
+impl<'a> PinDataMut<'a> {
+    fn new(data: &'a mut PinData) -> Self {
+        Self {
+            data,
+            modified: false,
+        }
+    }
+
+    pub fn check_pin(&mut self, pin: &Pin) -> bool {
+        if self.is_blocked() {
+            return false;
+        }
+        let success = hash(pin, &self.salt).ct_eq(&self.hash).into();
+        if let Some(retries) = &mut self.data.retries {
+            if success {
+                if retries.reset() {
+                    self.modified = true;
+                }
+            } else {
+                retries.decrement();
+                self.modified = true;
+            }
+        }
+        success
+    }
+}
+
+impl Deref for PinDataMut<'_> {
+    type Target = PinData;
+
+    fn deref(&self) -> &Self::Target {
+        self.data
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+struct Retries {
+    max: u8,
+    left: u8,
+}
+
+impl Retries {
+    fn decrement(&mut self) {
+        self.left = self.left.saturating_sub(1);
+    }
+
+    fn reset(&mut self) -> bool {
+        if self.left == self.max {
+            false
+        } else {
+            self.left = self.max;
+            true
+        }
+    }
+}
+
+impl From<u8> for Retries {
+    fn from(retries: u8) -> Self {
+        Self {
+            max: retries,
+            left: retries,
+        }
+    }
+}
+
+fn hash(pin: &Pin, salt: &Salt) -> Hash {
+    let mut digest = Sha256::new();
+    digest.update(pin);
```

Comment:
I would be in favour of adding `digest.update(&(pin.len() as u32).to_be_bytes())` to avoid any possible future canonicalization attack, an also integrate the PIN ID somehow.

Reply:
I’ll add it this evening.

Reply:
Done:
```rust
    digest.update([u8::from(id)]);
    digest.update([pin_len(pin)]);
```